# Problem : 02 : Gender recognition from Bangla handwritten digits

In this problem the **"Ekush"** dataset is used to recognize Male/Female from the Bangla handwritten digits. Deep neural network is used to build this binary classification model.

The Ekush dataset provides annoted data for the problem and has total of 30803 data samples. The training and testing samples are divided in **90:10** ratio. 

**Snapshot from Ekush**

<div align="center">
<img src="https://drive.google.com/uc?id=1GhuJfYchOjfNoBtwkpWivPRQinegMYwy" width="600">
</div>

- **Input dimension:**
  - Size of image: $28 \times 28 = 784$

- **Output dimension: 2**
  - 0, 1 [Female,Male]

For this particular problem we have used *different hyperparameters* like - batch size, number of iterations, learning rate , optimizer etc to get the best results. The combination of hyperparameters and their acquired accuracy is given below - 

| Hyperparameters| Setting 1| Setting 2  | Setting 3 | Setting 4  |
| :-------------:| :------: | :--------: | :-------: | :--------: | 
| Batch size     |    100   |    1024      |    512     |    128     |
| No. of iterations|  3000  |    3000    |    3000   |    3000    |
| Epoch          |    9.73  |    99.64    |    49.82   |    12.455   |
| Learning Rate  |    0.1 |    0.1   |    0.002  |    0.1   |
| Optimizer      |    SGD   |    SGD    |    SGD    |    SGD     |
| No of hidden Layers| 1  |    3     |    1    |    1    |
| No of hidden neurons  |    100   |    256,128    |    128   |    256     |
| Activation function|Sigmoid  |    ReLU,LeakyReLU,ReLU6     |    Tanh    |  ReLU     |
| ***Accuracy*** |   60.233 |    62.274  |    62.568 |    62.601  |

                                                    Table : Results of different hyperparameter setting

> From the table we can see the incremental results for each setting. The hyperparameters play an impactful role in the accuracy of model that is why tuning them right is an important thing to do. This experiment is done to get some basic intutives on the effects of hyperparameters. 


In **Setting 1** - I have taken the basic values of hyperparameters where the batch size is **100** and started the learning rate at **0.1**. in every setting **SGD** optimizer performs better or same as others that is why it is used in every setting. Only one hidden layer with 100 neurons and Sigmoid activation function is used here. The model achieves an accuracy of **60.233** with the **3000** iterations and **9.73** epochs. 

In **Setting 2** the **three** layer neural network is used where each layer has different number of neurons - **256->256->128** and different no of activation functions - **ReLU,LeakyReLU,ReLU6**. The model achieves an accuracy of **62.274** for this setting. 

However after experimenting for a while it is seen that *one* layer model gives better performance than *dense* model. So in **Setting-3** the hidden layer number is brought down to **one** and **Tanh** activation function is used with **128** neurons. This gives an accuracy of **62.568**. For this particular set of problem learning rate has shown no effect as **0.002** is used here but no promising increase is shown.

**Setting-4** shows that if we change the activation function to **ReLU** the accuracy is increased a little to **62.601**. Here only number of hidden neurons are increased to **256**. Among all the tried settings this setting gives us the best result. 

> After these experiments we can come to a conclusion that for this problem increasing the hidden layers or the batch size does not have big impact. Among them ReLU activation function works well. But this whole experiment gives us a conclusion that by perfectly tuning the hyperparameters binary classification model can give better results. Moreover, gender recognition is possible with Bangla handwritten digits. 

##Implementation : 

In [ ]:
import torch
import torch.nn as nn
import torch.nn.functional as F
import torchvision
import torchvision.transforms as transforms
import pandas as pd 
import numpy as np
import matplotlib.pyplot as plt 
from sklearn.model_selection import train_test_split
import PIL
from google.colab import drive
drive.mount('/drive')

Mounted at /drive


In [ ]:
df_female = pd.read_csv('/drive/My Drive/Colab Notebooks/SC_Lab_Assignments/femaleDigits.csv')
df_male = pd.read_csv('/drive/My Drive/Colab Notebooks/SC_Lab_Assignments/maleDigits.csv')
df_male.head()

pixel0  pixel1  pixel2  pixel3  ...  pixel781  pixel782  pixel783  label
0       0       2       0       1  ...         4         1         1    112
1       0       0       0       1  ...         1         0         0    112
2       1       0       0       0  ...         1         1         2    112
3       2       1       1       0  ...         0         0         2    112
4       0       0       0       0  ...         0         0         0    112

[5 rows x 785 columns]

In [ ]:
df_female['label'] = 'Female'
df_male['label'] = 'Male'
df_female.head()

pixel0  pixel1  pixel2  pixel3  ...  pixel781  pixel782  pixel783   label
0       0       0       0       0  ...         0         3         1  Female
1       0       0       0       0  ...         1         1         1  Female
2       0       0       1       0  ...         0         0         0  Female
3       0       0       0       0  ...         0         0         0  Female
4       2       3      12       5  ...         0         0         0  Female

[5 rows x 785 columns]

In [ ]:
dframes = [df_female, df_male]
dataset2 = pd.concat(dframes)
print(len(dataset2))

30830


In [ ]:
dataset2.head()

pixel0  pixel1  pixel2  pixel3  ...  pixel781  pixel782  pixel783   label
0       0       0       0       0  ...         0         3         1  Female
1       0       0       0       0  ...         1         1         1  Female
2       0       0       1       0  ...         0         0         0  Female
3       0       0       0       0  ...         0         0         0  Female
4       2       3      12       5  ...         0         0         0  Female

[5 rows x 785 columns]

In [ ]:
dataset2['label'] = dataset2['label'].astype('category')
dataset2.label = dataset2.label.cat.rename_categories([0,1]) #0-female , 1-male
#print (dataset2['label'].cat.categories)
dataset2.head()

pixel0  pixel1  pixel2  pixel3  ...  pixel781  pixel782  pixel783  label
0       0       0       0       0  ...         0         3         1      0
1       0       0       0       0  ...         1         1         1      0
2       0       0       1       0  ...         0         0         0      0
3       0       0       0       0  ...         0         0         0      0
4       2       3      12       5  ...         0         0         0      0

[5 rows x 785 columns]

In [ ]:
from sklearn.utils import shuffle
dataset2 = shuffle(dataset2)

In [ ]:
# Separating Data and Label
Y = dataset2["label"]
X = dataset2.drop(labels = ["label"],axis = 1)

# Normalize the data
X = X / 255.0

# Reshape the array into 28 x 28 pixel
X = X.values.reshape(-1,28,28,1)

In [ ]:
# split into train test sets
X_train, X_test, y_train, y_test = train_test_split(X, Y, test_size=0.1)
print(len(X_train))
print(len(y_train))
print(len(X_test))
print(len(y_test))

27747
27747
3083
3083


In [ ]:
train_dataset = list(zip(X_train,y_train))
test_dataset = list(zip(X_test,y_test))

### Setting-1  : 

- **totaldata:** 30830
- **minibatch:** 100
- **iterations:** 3,000
- **epochs**
  - $epochs = iterations \div \frac{totaldata}{minibatch} = 3000 \div \frac{30830}{100} = 9.73 $
- **Learning rate:** 0.1
- **Optimizer:** SGD
- **No of hidden Layers:** 1
- **No of hidden layer neurons:** 100
- **Activation Functions:** Sigmoid

In [ ]:
# Hyperparameters

batch_size = 100
num_iters = 3000
input_dim = 28*28 # num_features = 784
num_hidden = 100 # num of hidden nodes
output_dim = 2

learning_rate = 0.1  # More power so we can learn faster! previously it was 0.001

# Device
device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")

In [ ]:
'''
MAKING DATASET ITERABLE
'''
num_epochs = num_iters / (len(train_dataset) / batch_size)
num_epochs = int(num_epochs)

train_loader = torch.utils.data.DataLoader(dataset=train_dataset, 
                                           batch_size=batch_size, 
                                           shuffle=True)  

test_loader = torch.utils.data.DataLoader(dataset=test_dataset, 
                                          batch_size=batch_size, 
                                          shuffle=False)  

### Desigining model class

In [ ]:
class NeuralNetworkModel(nn.Module):
    def __init__(self, input_size, num_classes, num_hidden):
        super().__init__()
        ### 1st hidden layer
        self.linear_1 = nn.Linear(input_size, num_hidden)

        ### Non-linearity
        self.sigmoid = nn.Sigmoid()

        ### Output layer
        self.linear_out = nn.Linear(num_hidden, num_classes)

    def forward(self, x):
        # Linear layer
        out  = self.linear_1(x)
        # Non-linearity
        out = self.sigmoid(out)
        # Linear layer (output)
        probas  = self.linear_out(out)
        return probas

In [ ]:
'''
INSTANTIATE MODEL CLASS
'''
model = NeuralNetworkModel(input_size = input_dim,
                           num_classes = output_dim,
                           num_hidden = num_hidden)
# To enable GPU
model.to(device)

NeuralNetworkModel(
  (linear_1): Linear(in_features=784, out_features=100, bias=True)
  (sigmoid): Sigmoid()
  (linear_out): Linear(in_features=100, out_features=2, bias=True)
)

### Constructing loss and optimizer

In [ ]:
criterion = nn.CrossEntropyLoss()
optimizer = torch.optim.SGD(model.parameters(), lr=learning_rate)

### Training ,forward,loss,step

In [ ]:
'''
TRAIN THE MODEL
'''
iter = 0
for epoch in range(num_epochs):
    for i, (images, labels) in enumerate(train_loader):

        images = images.view(-1, 28*28).to(device)
        labels = labels.to(device)

        # Clear gradients w.r.t. parameters
        optimizer.zero_grad()

        # Forward pass to get output/logits
        outputs = model(images.float())

        # Calculate Loss: softmax --> cross entropy loss
        loss = criterion(outputs, labels)

        # Getting gradients w.r.t. parameters
        loss.backward()

        # Updating parameters
        optimizer.step()

        iter += 1

        if iter % 500 == 0:
            # Calculate Accuracy         
            correct = 0
            total = 0
            # Iterate through test dataset
            for images, labels in test_loader:
               
                images = images.view(-1, 28*28).to(device)

                # Forward pass only to get logits/output
                outputs = model(images.float())

                # Get predictions from the maximum value
                _, predicted = torch.max(outputs, 1)

                # Total number of labels
                total += labels.size(0)


                # Total correct predictions
                if torch.cuda.is_available():
                    correct += (predicted.cpu() == labels.cpu()).sum() 
                else:
                    correct += (predicted == labels).sum()

            accuracy = 100 * correct.item() / total

            # Print Loss
            print('Iteration: {}. Loss: {}. Accuracy: {}'.format(iter, loss.item(), accuracy))

Iteration: 500. Loss: 0.688839316368103. Accuracy: 59.876743431722346
Iteration: 1000. Loss: 0.7095091938972473. Accuracy: 60.13623094388583
Iteration: 1500. Loss: 0.6734268665313721. Accuracy: 59.29289652935452
Iteration: 2000. Loss: 0.6247273087501526. Accuracy: 59.03340901719105
Iteration: 2500. Loss: 0.6869900226593018. Accuracy: 60.23353876094713


### Setting-2  : 

- **totaldata:** 30830
- **minibatch:** 1024
- **iterations:** 3,000
- **epochs**
  - $epochs = iterations \div \frac{totaldata}{minibatch} = 3000 \div \frac{30830}{1024} = 99.64 $
- **Learning rate:** 0.1
- **Optimizer:** SGD
- **No of hidden Layers:** 3
- **No of hidden layer neurons:** 256->128
- **Activation Functions:** ReLU->LeakyReLU->ReLU6

In [ ]:
import torch
import torch.nn as nn
import torchvision.transforms as transforms
import torchvision.datasets as dsets

# Hyperparameters
batch_size = 1024
num_iters = 3000
input_dim = 28*28 #num_features = 784
num_hidden = 256
output_dim = 2

learning_rate = 0.1

# Device
device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")

from sklearn.utils import shuffle
dataset2 = shuffle(dataset2)
# Separating Data and Label
Y = dataset2["label"]
X = dataset2.drop(labels = ["label"],axis = 1)

# Normalize the data
X = X / 255.0

# Reshape the array into 28 x 28 pixel
X = X.values.reshape(-1,28,28,1)
X_train, X_test, y_train, y_test = train_test_split(X, Y, test_size=0.1)

train_dataset = list(zip(X_train,y_train))
test_dataset = list(zip(X_test,y_test))

num_epochs = num_iters / (len(train_dataset) / batch_size)
num_epochs = int(num_epochs)

train_loader = torch.utils.data.DataLoader(dataset=train_dataset, 
                                           batch_size=batch_size, 
                                           shuffle=True)   # It's better to shuffle the whole training dataset! 

test_loader = torch.utils.data.DataLoader(dataset=test_dataset, 
                                          batch_size=batch_size, 
                                          shuffle=False) 

class DeepNeuralNetworkModel(nn.Module):
    def __init__(self, input_size, num_classes, num_hidden):
        super().__init__()
        ### 1st hidden layer: 784 --> 100
        self.linear_1 = nn.Linear(input_size, num_hidden)
        ### Non-linearity in 1st hidden layer
        self.relu_1 = nn.ReLU()

        ### 2nd hidden layer: 100 --> 100
        self.linear_2 = nn.Linear(num_hidden, num_hidden)
        ### Non-linearity in 2nd hidden layer
        self.relu_2 = nn.LeakyReLU()

        ### 3rd hidden layer: 100 --> 100
        self.linear_3 = nn.Linear(num_hidden, 128)
        ### Non-linearity in 3rd hidden layer
        self.relu_3 = nn.ReLU6()

        ### Output layer: 100 --> 10
        self.linear_out = nn.Linear(128, num_classes)

    def forward(self, x):
        ### 1st hidden layer
        out  = self.linear_1(x)
        ### Non-linearity in 1st hidden layer
        out = self.relu_1(out)
        
        ### 2nd hidden layer
        out  = self.linear_2(out)
        ### Non-linearity in 2nd hidden layer
        out = self.relu_2(out)

        ### 3rd hidden layer
        out  = self.linear_3(out)
        ### Non-linearity in 3rd hidden layer
        out = self.relu_3(out)
        
        # Linear layer (output)
        probas  = self.linear_out(out)
        return probas

# INSTANTIATE MODEL CLASS

model = DeepNeuralNetworkModel(input_size = input_dim,
                               num_classes = output_dim,
                               num_hidden = num_hidden)
# To enable GPU
model.to(device)

# INSTANTIATE LOSS & OPTIMIZER CLASS
criterion = nn.CrossEntropyLoss()
optimizer = torch.optim.SGD(model.parameters(), lr=learning_rate)

iter = 0
for epoch in range(num_epochs):
    for i, (images, labels) in enumerate(train_loader):

        images = images.view(-1, 28*28).to(device)
        labels = labels.to(device)

        # Clear gradients w.r.t. parameters
        optimizer.zero_grad()

        # Forward pass to get output/logits
        outputs = model(images.float()) 

        # Calculate Loss: softmax --> cross entropy loss
        loss = criterion(outputs, labels)

        # Getting gradients w.r.t. parameters
        loss.backward()

        # Updating parameters
        optimizer.step()

        iter += 1

        if iter % 500 == 0:
            # Calculate Accuracy         
            correct = 0
            total = 0
            # Iterate through test dataset
            for images, labels in test_loader:
               
                images = images.view(-1, 28*28).to(device)

                # Forward pass only to get logits/output
                outputs = model(images.float())

                # Get predictions from the maximum value
                _, predicted = torch.max(outputs, 1)

                # Total number of labels
                total += labels.size(0)


                # Total correct predictions
                if torch.cuda.is_available():
                    correct += (predicted.cpu() == labels.cpu()).sum() 
                else:
                    correct += (predicted == labels).sum()

            accuracy = 100 * correct.item() / total

            # Print Loss
            print('Iteration: {}. Loss: {}. Accuracy: {}'.format(iter, loss.item(), accuracy))

Iteration: 500. Loss: 0.6453047394752502. Accuracy: 58.90366526110931
Iteration: 1000. Loss: 0.6199793219566345. Accuracy: 58.77392150502757
Iteration: 1500. Loss: 0.6013468503952026. Accuracy: 62.504054492377556
Iteration: 2000. Loss: 0.5936097502708435. Accuracy: 62.76354200454103
Iteration: 2500. Loss: 0.5891481041908264. Accuracy: 62.37431073629582


### Setting-3  : 

- **totaldata:** 30830
- **minibatch:** 512
- **iterations:** 3,000
- **epochs**
  - $epochs = iterations \div \frac{totaldata}{minibatch} = 3000 \div \frac{30830}{512} = 49.82 $
- **Learning rate:** 0.002
- **Optimizer:** SGD
- **No of hidden Layers:** 1
- **No of hidden layer neurons:** 128
- **Activation Functions:** Tanh

In [ ]:
import torch
import torch.nn as nn
import torchvision.transforms as transforms
import torchvision.datasets as dsets

# Hyperparameters
batch_size = 512
num_iters = 3000
input_dim = 28*28 # num_features = 784
num_hidden = 128
output_dim = 2

learning_rate = 0.1

# Device
device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")


num_epochs = num_iters / (len(train_dataset) / batch_size)
num_epochs = int(num_epochs)

from sklearn.utils import shuffle
dataset2 = shuffle(dataset2)
# Separating Data and Label
Y = dataset2["label"]
X = dataset2.drop(labels = ["label"],axis = 1)

# Normalize the data
X = X / 255.0

# Reshape the array into 28 x 28 pixel
X = X.values.reshape(-1,28,28,1)
X_train, X_test, y_train, y_test = train_test_split(X, Y, test_size=0.1)

train_dataset = list(zip(X_train,y_train))
test_dataset = list(zip(X_test,y_test))



train_loader = torch.utils.data.DataLoader(dataset=train_dataset, 
                                           batch_size=batch_size, 
                                           shuffle=True)  

test_loader = torch.utils.data.DataLoader(dataset=test_dataset, 
                                          batch_size=batch_size, 
                                          shuffle=False)  


class NeuralNetworkModel(nn.Module):
    def __init__(self, input_size, num_classes, num_hidden):
        super().__init__()
        ### 1st hidden layer
        self.linear_1 = nn.Linear(input_size, num_hidden)

        ### Non-linearity
        self.tanh = nn.Tanh()

        ### Output layer
        self.linear_out = nn.Linear(num_hidden, num_classes)

    def forward(self, x):
        # Linear layer
        out  = self.linear_1(x)
        # Non-linearity
        out = self.tanh(out)
        # Linear layer (output)
        probas  = self.linear_out(out)
        return probas

model = NeuralNetworkModel(input_size = input_dim,
                           num_classes = output_dim,
                           num_hidden = num_hidden)
# To enable GPU
model.to(device)

criterion = nn.CrossEntropyLoss()
optimizer = torch.optim.SGD(model.parameters(), lr=learning_rate)

iter = 0
for epoch in range(num_epochs):
    for i, (images, labels) in enumerate(train_loader):

        images = images.view(-1, 28*28).to(device)
        labels = labels.to(device)

        # Clear gradients w.r.t. parameters
        optimizer.zero_grad()

        # Forward pass to get output/logits
        outputs = model(images.float())

        # Calculate Loss: softmax --> cross entropy loss
        loss = criterion(outputs, labels)

        # Getting gradients w.r.t. parameters
        loss.backward()

        # Updating parameters
        optimizer.step()

        iter += 1

        if iter % 500 == 0:
            # Calculate Accuracy         
            correct = 0
            total = 0
            # Iterate through test dataset
            for images, labels in test_loader:
               
                images = images.view(-1, 28*28).to(device)

                # Forward pass only to get logits/output
                outputs = model(images.float())

                # Get predictions from the maximum value
                _, predicted = torch.max(outputs, 1)

                # Total number of labels
                total += labels.size(0)


                # Total correct predictions
                if torch.cuda.is_available():
                    correct += (predicted.cpu() == labels.cpu()).sum() 
                else:
                    correct += (predicted == labels).sum()

            accuracy = 100 * correct.item() / total

            # Print Loss
            print('Iteration: {}. Loss: {}. Accuracy: {}'.format(iter, loss.item(), accuracy))

Iteration: 500. Loss: 0.6452792286872864. Accuracy: 61.401232565682776
Iteration: 1000. Loss: 0.6445723176002502. Accuracy: 61.17418099253973
Iteration: 1500. Loss: 0.6344114542007446. Accuracy: 62.504054492377556
Iteration: 2000. Loss: 0.6264241337776184. Accuracy: 62.27700291923451
Iteration: 2500. Loss: 0.6586355566978455. Accuracy: 62.568926370418424


### Setting-4  : 

- **totaldata:** 30830
- **minibatch:** 128
- **iterations:** 3,000
- **epochs**
  - $epochs = iterations \div \frac{totaldata}{minibatch} = 3000 \div \frac{30830}{128} = 12.455 $
- **Learning rate:** 0.1
- **Optimizer:** SGD
- **No of hidden Layers:** 1
- **No of hidden layer neurons:** 256
- **Activation Functions:** ReLU

In [ ]:
import torch
import torch.nn as nn
import torchvision.transforms as transforms
import torchvision.datasets as dsets

# Hyperparameters
batch_size = 128
num_iters = 3000
input_dim = 28*28 # num_features = 784
num_hidden = 256
output_dim = 2

learning_rate = 0.1

# Device
device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")

from sklearn.utils import shuffle
dataset2 = shuffle(dataset2)
# Separating Data and Label
Y = dataset2["label"]
X = dataset2.drop(labels = ["label"],axis = 1)

# Normalize the data
X = X / 255.0

# Reshape the array into 28 x 28 pixel
X = X.values.reshape(-1,28,28,1)
X_train, X_test, y_train, y_test = train_test_split(X, Y, test_size=0.1)

train_dataset = list(zip(X_train,y_train))
test_dataset = list(zip(X_test,y_test))

num_epochs = num_iters / (len(train_dataset) / batch_size)
num_epochs = int(num_epochs)

train_loader = torch.utils.data.DataLoader(dataset=train_dataset, 
                                           batch_size=batch_size, 
                                           shuffle=True)   # It's better to shuffle the whole training dataset! 

test_loader = torch.utils.data.DataLoader(dataset=test_dataset, 
                                          batch_size=batch_size, 
                                          shuffle=False) 

class NeuralNetworkModel(nn.Module):
    def __init__(self, input_size, num_classes, num_hidden):
        super().__init__()
        ### 1st hidden layer
        self.linear_1 = nn.Linear(input_size, num_hidden)

        ### Non-linearity
        self.relu = nn.ReLU()

        ### Output layer
        self.linear_out = nn.Linear(num_hidden, num_classes)

    def forward(self, x):
        # Linear layer
        out  = self.linear_1(x)
        # Non-linearity
        out = self.relu(out)
        # Linear layer (output)
        probas  = self.linear_out(out)
        return probas

model = NeuralNetworkModel(input_size = input_dim,
                           num_classes = output_dim,
                           num_hidden = num_hidden)
# To enable GPU
model.to(device)

criterion = nn.CrossEntropyLoss()
optimizer = torch.optim.SGD(model.parameters(), lr=learning_rate)

iter = 0
for epoch in range(num_epochs):
    for i, (images, labels) in enumerate(train_loader):

        images = images.view(-1, 28*28).to(device)
        labels = labels.to(device)

        # Clear gradients w.r.t. parameters
        optimizer.zero_grad()

        # Forward pass to get output/logits
        outputs = model(images.float()) 

        # Calculate Loss: softmax --> cross entropy loss
        loss = criterion(outputs, labels)

        # Getting gradients w.r.t. parameters
        loss.backward()

        # Updating parameters
        optimizer.step()

        iter += 1

        if iter % 500 == 0:
            # Calculate Accuracy         
            correct = 0
            total = 0
            # Iterate through test dataset
            for images, labels in test_loader:
               
                images = images.view(-1, 28*28).to(device)

                # Forward pass only to get logits/output
                outputs = model(images.float())

                # Get predictions from the maximum value
                _, predicted = torch.max(outputs, 1)

                # Total number of labels
                total += labels.size(0)


                # Total correct predictions
                if torch.cuda.is_available():
                    correct += (predicted.cpu() == labels.cpu()).sum() 
                else:
                    correct += (predicted == labels).sum()

            accuracy = 100 * correct.item() / total

            # Print Loss
            print('Iteration: {}. Loss: {}. Accuracy: {}'.format(iter, loss.item(), accuracy))

Iteration: 500. Loss: 0.6357078552246094. Accuracy: 62.66623418747973
Iteration: 1000. Loss: 0.6330263614654541. Accuracy: 62.504054492377556
Iteration: 1500. Loss: 0.5976560115814209. Accuracy: 62.568926370418424
Iteration: 2000. Loss: 0.611310601234436. Accuracy: 62.244566980214074
Iteration: 2500. Loss: 0.6447990536689758. Accuracy: 62.601362309438855
